beacon-chain/blockchain/service.go

```diff
@@ -299,6 +405,46 @@ func (c *ChainService) processBlock(block *types.Block) error {
 	return nil
 }
 
+func (c *ChainService) processBlockNew(block *types.Block) error {
+
+	blockhash, err := block.Hash()
+	if err != nil {
+		return fmt.Errorf("could not hash incoming block: %v", err)
+	}
+
+	beaconState, err := c.beaconDB.GetState()
+	if err != nil {
+		return fmt.Errorf("failed to get beacon state: %v", err)
+	}
+
+	if block.SlotNumber() == 0 {
+		return errors.New("cannot process a genesis block: received block with slot 0")
+	}
+
+	if !c.isBlockReadyForProcessing(block) {
+		c.unProcessedBlocks[block.SlotNumber()] = block
+		return nil
+	}
+
+	// Add in slot transition method here, where signatures,randao and
+	// block processing will be validated.
+
+	// Add in epoch transition method here, where votes, validator exits,
+	// and balances will be updated.
```

Comment:
This should simply follow the pseudocode outlined by Danny [here](https://notes.ethereum.org/s/BJEZWNoyE) as:
```python
def process_block(block):
    if not block_pre_processing_conditions(block):
        return False
        
    # process skipped slots 
    while (state.slot < block.slot - 1):
        state = slot_state_transition(state, block=None)
        
    # process slot with block
    state = slot_state_transition(state, block)
    
    # check state root
    if block.state_root == hash(state):
        return state
    else:
        return False  # or throw or whatever
```
We shouldn't need two separate state transitions, just a single state transition function that can check if block != nil should do it.

Reply:
I wanted to hold off on writing this part into our codebase: 
```py
    # process skipped slots 
    while (state.slot < block.slot - 1):
        state = slot_state_transition(state, block=None)
```
This assumes the blocks all come in order sequentially, while there is a good chance that might not be the case. Danny is going to be working on some notes to illustrate how to handle the edge cases